datameta/api/metadatasets.py

```diff
@@ -348,3 +347,95 @@ def delete_metadataset(request:Request) -> HTTPNoContent:
     delete_staged_metadataset_from_db(request.matchdict['id'], db, auth_user, request)
 
     return HTTPNoContent()
+
+
+@view_config(
+    route_name="service_execution",
+    renderer='json',
+    request_method="POST",
+    openapi=True
+)
+def set_metadata_via_service(request:Request) -> MetaDataSetResponse:
+    """Endpoint for the submission of results of a service execution."""
+
+    # Check authentication or raise 401
+    auth_user = security.revalidate_user(request)
+    db = request.dbsession
+
+    # Query the specified resources
+    metadataset = resource.resource_query_by_id(db, MetaDataSet, request.matchdict['metadatasetId'])\
+            .options(joinedload(MetaDataSet.service_executions).joinedload(ServiceExecution.service))\
+            .options(joinedload(MetaDataSet.metadatumrecords).joinedload(MetaDatumRecord.metadatum))\
+            .one_or_none()
+
+    service = resource.resource_query_by_id(db, Service, request.matchdict['serviceId'])\
+            .options(joinedload(Service.target_metadata))\
+            .options(joinedload(Service.users))\
+            .one_or_none()
+
+    service_metadata = { mdatum.name : mdatum for mdatum in service.target_metadata }
+
+    service_records = { rec.metadatum.name : rec for rec in metadataset.metadatumrecords if rec.metadatum.name in service_metadata }
+
+    # Return 404 if a resource could not be found
+    if metadataset is None or service is None:
+        raise HTTPNotFound()
+
+    # Return 403 if the user has no permission to execute this service or the
+    # service has already been executed for this metadataset
+    if service.id in (sexec.service_id for sexec in metadataset.service_executions):
+        raise HTTPForbidden(json_body={})
+
+    if not authz.execute_service(auth_user, service):
+        raise HTTPForbidden(json_body={})
+
+    # Try to associate all metadata in the request body with the service metadata
+    records = record_to_strings(request.openapi_validated.body["record"])
+    val_errors = [] # tuples (message, field)
+    for mdatum_name in records:
+        if mdatum_name not in (target_metadatum.name for target_metadatum in service.target_metadata):
+            val_errors.append(("Metadatum unknown or not associated with the specified service.", mdatum_name))
```

Comment:
```suggestion
    service_assoc_metadata = (target_metadatum.name for target_metadatum in service.target_metadata)
    err_msg = "Metadatum unknown or not associated with the specified service."
    val_errors = [(err_msg, mdatum_name) for mdatum_name in set(records).difference(service_assoc_metadata)]
```

My idea is to avoid recreating the `(target_metadatum.name for target_metadatum in service.target_metadata) generator` for each metadatum in records.

Reply:
Interesting solution, I'm not quite sure about the performance gain though. I prefer the old code for readability.